### DISCUSSION TITLE
[Http-Client] Concurrent Requests total time != longest response

### DISCUSSION THREAD

**Discussion by DoctorBooooom** [symfony/symfony#52527] [Q&A]
I was trying to check how much faster parallel requests work than if you request them sequentially. And when using asynchronous requests, I expected that the total waiting time would be approximately equal to the longest response. However, the total time, although less, is not equal to the longest answer.
Can someone explain why?

I also tried swapping queries so that the longest ones were the first, but it didn't work.

Synchronous requests:
```php
$this->httpClient->request('GET', 'https://........../dictionary/country', ['verify_host' => false, 'verify_peer' => false])->getContent();
$this->httpClient->request('GET', 'https://........../stories/feed', ['verify_host' => false, 'verify_peer' => false])->getContent();
$this->httpClient->request('GET', 'https://........../sms/empty', ['verify_host' => false, 'verify_peer' => false])->getContent();
$this->httpClient->request('GET', 'https://........../dictionary/program', ['verify_host' => false, 'verify_peer' => false])->getContent();
$this->httpClient->request('GET', 'https://........../mobile/country', ['verify_host' => false, 'verify_peer' => false])->getContent();
$this->httpClient->request('GET', 'https://........../lessons', ['verify_host' => false, 'verify_peer' => false])->getContent();
$this->httpClient->request('GET', 'https://........../help-actions', ['verify_host' => false, 'verify_peer' => false])->getContent();
$this->httpClient->request('GET', 'https://........../cities', ['verify_host' => false, 'verify_peer' => false])->getContent();
```
![image](https://github.com/symfony/symfony/assets/3152959/e0da6566-ed75-47eb-b8a7-52ed708681f4)

Asynchronous requests:
```php
$responseList[] = $this->httpClient->request('GET', 'https://........../dictionary/country', ['verify_host' => false, 'verify_peer' => false]);
$responseList[] = $this->httpClient->request('GET', 'https://........../stories/feed', ['verify_host' => false, 'verify_peer' => false]);
$responseList[] = $this->httpClient->request('GET', 'https://........../sms/empty', ['verify_host' => false, 'verify_peer' => false]);
$responseList[] = $this->httpClient->request('GET', 'https://........../dictionary/program', ['verify_host' => false, 'verify_peer' => false]);
$responseList[] = $this->httpClient->request('GET', 'https://........../mobile/country', ['verify_host' => false, 'verify_peer' => false]);
$responseList[] = $this->httpClient->request('GET', 'https://........../lessons', ['verify_host' => false, 'verify_peer' => false]);
$responseList[] = $this->httpClient->request('GET', 'https://........../help-actions', ['verify_host' => false, 'verify_peer' => false]);
$responseList[] = $this->httpClient->request('GET', 'https://........../cities', ['verify_host' => false, 'verify_peer' => false]);

foreach ($responseList as $response) {
    $response->getContent();
}
```
![image](https://github.com/symfony/symfony/assets/3152959/8b4534e7-b2ef-4ccc-a6bf-a1c17cc31137)

**Comment by stof** [comment]
Well, the traceable client will register the end of the request when the client reads that response for it.
In your code, the network IO is concurrent. But your loop waiting for response content is still sequential, where you process the responses in the order in which you defined them, not in the order in which they arrive.
If you want to process each response as soon as it arrives, you will need to use the `$client->stream()` method instead of looping over responses in a fixed order and using the blocking method.

Regarding the total waiting time not being the time of the slowest query, maybe you saturate your network bandwidth and so concurrent requests are slower than each request done alone because it has less bandwidth available (note that the total time is still improved, because the bandwidth was not the _only_  limiting factor, there is also the server processing time that impacts the query time)
